Data/IntMap/Internal.hs

```diff
@@ -399,7 +399,9 @@ bitmapOf x = shiftLL 1 (x .&. IntSet.suffixBitMask)
 #else
 (!) :: IntMap a -> Key -> a
 #endif
-(!) m k = find k m
+(!) m k
+  | Just a <- lookup k m = a
+  | otherwise = error ("IntMap.!: key " ++ show k ++ " is not an element of the map")
```

Comment:
Using `lookup` will incur a `Just` allocation. One possible solution (for very recent GHC versions only) would be to define a `lookup#` function producing an unboxed sum:

```haskell
type Maybe# a = (# Void# | a #)

pattern Nothing# :: Maybe# a
pattern Nothing# <- (# _ | #) where
  Nothing# = (# void# | #)

pattern Just# :: a -> Maybe# a
pattern Just# a = (# | a #)

lookup# :: Key -> IntMap a -> Maybe# a
lookup# k m = ...

lookup k m = case lookup# k m of
  Nothing# -> Nothing
  Just# a -> Just a
```

The (potentially) great thing about this is that the `Just` gets applied on the "outside", so inlining and the case-of-case transformation will end up making it go away altogether:

```haskell
m ! k = case
    (case lookup# k m of
       Nothing# -> Nothing
       Just# a -> Just a) of
       Nothing -> error ...
       Just v -> v

-- ==> case-of-case, case of known constructor

m ! k =
  case lookup# k m of
       Nothing# -> error ...
       Just# a -> a
```

Unlike a CPS version, this theoretically helps user-written code as well. However, I have not actually tried benchmarking anything like this. And the CPP is potentially troublesome too.

Reply:
This seems cool -- it seems like it could be a fairly big change that would even prevent an allocation when one uses `lookup` (or related functions that return `Maybe` values) and immediately unwraps -- but it's a bit big for what I was hoping to contribute just now (also, a bit over my head).  I think I'll take the easy way out and go back to using a `find` function to get a result without the extra allocation.  Perhaps my next PR will be to overhaul the backend to use unboxed sums where possible :).